### PR TITLE
Fix for iontorrent uploads

### PR DESCRIPTION
**Pull request by CEJ-R**
### Priority: 
(High)

### What does this PR aim to solve:
There appears to have been some problems with the uploads of samples from the iontorrent machines. To solve this the code handling this will need to be changed and reviewed. One potential issue are the various versions of pangolin, and how they might affect the parsing.

### Changed in this PR:
- torrent-plugin
  - Different handling of csv-separators for pangolin
  - Added better README
  - More descriptive messages in stdout
  - Changed which header determines pangolin fail
  - Changed emails to cgg-micro where found
  - Now doesn't exclude headers from metadata (should never be there...)
- server-side
  - Handling if samples are included in metadata, but didn't generate/tranfer any files
  - Determination if pangolin 4
  - Check if files are uploaded and add to email
  - Moved emails to config for ease of editing

### CODE REVIEW

File: microbiology-plugin/covid_seqstore_transfer/covid_seqstore_transfer.py [hunk 1]
```diff
@@ -108,7 +108,7 @@ def parse_pangolin_csv(pangolin_csv_path):
     """Parse csv into more fun to read format."""
     collected_info = {}
     with open(pangolin_csv_path, 'r') as csvfile:
-        dialect = csv.Sniffer().sniff(csvfile.read(), delimiters=';,\t')
+        dialect = csv.Sniffer().sniff(csvfile.readline(), [',',';','\t'])
```
Comment (JPWerling) [inline microbiology-plugin/covid_seqstore_transfer/covid_seqstore_transfer.py:111]: missing keyword `delimiters=`

Comment (JPWerling) [inline microbiology-plugin/covid_seqstore_transfer/covid_seqstore_transfer.py:111]: `Sniff` does not work for tabs
https://stackoverflow.com/questions/51074335/delimiter-with-tab-in-text-file

**Review by JPWerling — CHANGES_REQUESTED** [review]

**Comment by CEJ-R** [comment]
@JPWerling Then I'm not sure what to say, as it seems to work fine when run on the iontorrent (just this part of the script) `/home/ionadmin/misc_scripts/emilio_covid_script/broken_covid_seqstore_transfer.py`. There I have this function, fed with some xls', and it seems to produce a reasonable dict when written like the PR currently has it.

**Comment by CEJ-R** [comment]
This should now work as intended and is ready for merging before I deal with the wrapper-side.

File: microbiology_s5/runner.py [hunk 2]
```diff
@@ -100,9 +121,16 @@ def run_single(run_path, working_directory_root, log_path, cleanup, test):
 
             # All uploads here
             for f_path in to_upload:
+                filename = os.path.basename(f_path)
                 if not test:
-                    logger.info(f'Uploading file: {f_path}')
-                    sftp_connection.put(f_path, os.path.basename(f_path))
+                    logger.info(f'Uploading file: {filename}')
+                    sftp_connection.put(f_path, os.path.basename(filename))
```
Comment (JPWerling) [inline microbiology_s5/runner.py:127]: Redundant `basename`

File: microbiology_s5/src/reformat.py [hunk 3]
```diff
@@ -41,15 +45,23 @@ def reformat_pangolin(pangolin_path, output_path):
...
         headers = ['provnummer', 'urvalskriterium', 'GISAID_accession']
         print(','.join(headers), file=out)
 
         for sample_name, sample_metadata in metadata.items():
-            selection_criteria_index = int(sample_metadata.split('_')[-1])
-            selection_criteria_text = selection_criteria[selection_criteria_index]
-            print(','.join([sample_name, selection_criteria_text, '']), file=out)
+            any_missed = [a for a in missing if sample_name in a]  # Parse any missing files for this sample
+            if len(any_missed) != 3:  # If this sample had any uploaded files, add to complementary
+                clean_meta = sample_metadata.rstrip()  # Clean newline from string
```
Comment (JPWerling) [inline microbiology_s5/src/reformat.py:57]: could fix this in the js parser

File: microbiology_s5/src/reformat.py [hunk 4]
```diff
@@ -41,15 +45,23 @@ def reformat_pangolin(pangolin_path, output_path):
...
+                # Handle possibility of samples missing criterion field
+                if len(clean_meta.split('_')) != 15 or clean_meta.split('_')[-1] == "null":
+                    selection_criteria_text = 'Information saknas'
+                else:
+                    selection_criteria_index = int(clean_meta.split('_')[-1])
+                    selection_criteria_text = selection_criteria[selection_criteria_index]
```
Comment (JPWerling) [inline microbiology_s5/src/reformat.py:63]: I don't see `selection_criteria` defined

File: microbiology_s5/src/reformat.py [hunk 5]
```diff
@@ -41,15 +45,23 @@ def reformat_pangolin(pangolin_path, output_path):
...
+
+                print(','.join([sample_name, selection_criteria_text, '']), file=out)
```
Comment (JPWerling) [inline microbiology_s5/src/reformat.py:65]: Could use a comment on what the last `''` is for

**Review by JPWerling — CHANGES_REQUESTED** [review]
Everything looks to function apart from a missing `selection_criteria` definition unless I'm blind

File: microbiology-plugin/covid_seqstore_transfer/covid_seqstore_transfer.py [hunk 6]
```diff
@@ -192,7 +192,7 @@ def launch(self):
             except KeyError:
                 no_metadata_samples.add_sample(barcode, sample_name)
                 covid_samples.remove_sample(barcode)
-                print("barcode: {} did not have metadata".format(barcode))
+                print("barcode: {} did not have metadata:".format(barcode)) # Note that the top row of metadata is assumed to be header
```
[the remark this answers is not in the document]
Comment (CEJ-R) [reply microbiology-plugin/covid_seqstore_transfer/covid_seqstore_transfer.py]: Yes... this was before I knew that there would (probably) never be headers. Changed!

File: microbiology_s5/config.py [hunk 7]
```diff
@@ -7,6 +7,8 @@
 
 ADMIN_EMAIL = 'cgg-micro@gu.se'
 GENSAM_EMAIL = 'gensam@folkhalsomyndigheten.se'
+SUMMARY_EMAIL = ['cgg-micro@gu.se']
+ERROR_EMAIL = ['cgg-micro@gu.se']
```
[the remark this answers is not in the document]
Comment (CEJ-R) [reply microbiology_s5/config.py:11]: I think this is probably better as separate list. KMIK usually want different people to get different sendouts with the covid-stuff. I do intend on adding a bunch more emails here once this goes live.

File: microbiology_s5/runner.py (hunk 2 above)
Comment (CEJ-R) [reply microbiology_s5/runner.py]: Good catch!

File: microbiology_s5/src/reformat.py (hunk 4 above)
Comment (CEJ-R) [reply microbiology_s5/src/reformat.py:63]: It reads from an import:
https://github.com/ClinicalGenomicsGBG/sars-cov-2-typing/blob/232d958feb4ac00d31d2decd22b62c8e1726d8f2/microbiology_s5/src/reformat.py#L9

File: microbiology_s5/src/reformat.py (hunk 5 above)
Comment (CEJ-R) [reply microbiology_s5/src/reformat.py]: Yes... and this needs patching later as well. More covid fun for the future!

File: microbiology_s5/src/reformat.py (hunk 3 above)
Comment (CEJ-R) [reply microbiology_s5/src/reformat.py:57]: I think I'll leave it as is here (python), since I know how this is supposed to work.

**Review by JPWerling — APPROVED** [review]

**Pull request merged**